modules/local-caching-classloader/src/main/java/org/apache/accumulo/classloader/lcc/LocalCachingContextClassLoaderFactory.java

```diff
@@ -0,0 +1,193 @@
+ * this class should monitor the file for changes and a list of {@link Resource} objects. Each
+ * resource is defined by a URL to the file and an expected MD5 hash value.
+ * <p>
+ * The URLs supplied for the context definition file and for the resources can use one of the
+ * following protocols: file://, http://, or hdfs://.
+ * <p>
+ * As this class processes the ContextDefinition it fetches the contents of the resource from the
+ * resource URL and caches it in a directory on the local filesystem. This class uses the value of
+ * the system property {@link Constants#CACHE_DIR_PROPERTY} as the root directory and creates a
+ * sub-directory for each context name. Each context cache directory contains a lock file and a copy
+ * of each fetched resource that is named using the following format: fileName_checksum.
+ * <p>
+ * The lock file prevents processes from manipulating the contexts of the context cache directory
+ * concurrently, which enables the cache directories to be shared among multiple processes on the
+ * host.
+ * <p>
+ * Note that because the cache directory is shared among multiple processes, and one process can't
+ * know what the other processes are doing, this class cannot clean up the shared cache directory.
+ * It is left to the user to remove unused context cache directories and unused old files within a
+ * context cache directory.
+ */
+public class LocalCachingContextClassLoaderFactory implements ContextClassLoaderFactory {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(LocalCachingContextClassLoaderFactory.class);
+
+  private final Cache<String,LocalCachingContext> contexts =
+      Caffeine.newBuilder().expireAfterAccess(24, TimeUnit.HOURS).build();
+
+  private ContextDefinition parseContextDefinition(final URL url)
+      throws ContextClassLoaderException {
+    LOG.trace("Retrieving context definition file from {}", url);
+    final FileResolver resolver = FileResolver.resolve(url);
+    try {
+      try (InputStream is = resolver.getInputStream()) {
+        ContextDefinition def =
+            Constants.GSON.fromJson(new InputStreamReader(is, UTF_8), ContextDefinition.class);
+        if (def == null) {
+          throw new ContextClassLoaderException(
+              "ContextDefinition null for context definition file: " + resolver.getURL());
+        }
+        return def;
+      }
+    } catch (IOException e) {
+      throw new ContextClassLoaderException(
+          "Error reading context definition file: " + resolver.getURL(), e);
+    }
+  }
+
+  /**
+   * Schedule a task to execute at {@code interval} seconds to update the LocalCachingContext if the
+   * ContextDefinition has changed. The task schedules a follow-on task at the update interval value
+   * (if it changed).
+   */
+  private void monitorContext(final String contextLocation, int interval) {
+    Constants.EXECUTOR.schedule(() -> {
+      final LocalCachingContext classLoader = contexts.getIfPresent(contextLocation);
```

Comment:
We discussed that this cache access by the background refresh task could keep the cache entry from ever expiring. Need to avoid accessing the cache here.  This could be a follow on issue.  We discussed using a weaf ref or the cache eviction notification mechanism as ways to avoid this cache access.